DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs

```diff
@@ -94,5 +115,362 @@ public IActionResult AllCourseStatistics()
+                    (s, index) =>
+                    {
+                        var tutorials = tutorialService.GetTutorialsForSection(s.Id);
+                        return new SetSectionContentViewModel(s, index, data!.Application!.DiagAssess, tutorials);
+                    }
+                );
+                data!.SetSectionContentModels = availableSections.ToList();
+            }
+
+            if (!ModelState.IsValid)
+            {
+                return View("AddNewCentreCourse/SetCourseContent", model);
+            }
+
+            data!.SetCourseContentModel = model;
+            TempData.Set(data);
+
+            return RedirectToAction(model.IncludeAllSections ? "Summary" : "SetSectionContent");
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpGet]
+        [Route("AddCourse/SetSectionContent")]
+        public IActionResult SetSectionContent(int sectionIndex)
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>();
+
+            var section = data!.SetCourseContentModel!.GetSelectedSections().ElementAt(sectionIndex);
+            var tutorials = tutorialService.GetTutorialsForSection(section.Id).ToList();
+            var showDiagnostic = data.Application!.DiagAssess;
+            var model = new SetSectionContentViewModel(section, sectionIndex, showDiagnostic, tutorials);
+
+            return !tutorials.Any()
+                ? RedirectToNextSectionOrSummary(model.Index, data.SetCourseContentModel)
+                : View("AddNewCentreCourse/SetSectionContent", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpPost]
+        [Route("AddCourse/SetSectionContent")]
+        public IActionResult SetSectionContent(
+            SetSectionContentViewModel model,
+            string action
+        )
+        {
+            if (action == SaveAction)
+            {
+                return SaveSectionAndRedirect(model);
+            }
+
+            var bulkSelectResult = EditCourseSectionHelper.ProcessBulkSelect(model, action);
+            return bulkSelectResult ?? View("AddNewCentreCourse/SetSectionContent", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpGet]
+        [Route("AddCourse/Summary")]
+        public IActionResult Summary()
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>();
+
+            var model = new SummaryViewModel(data!);
+
+            return View("AddNewCentreCourse/Summary", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpPost]
+        [Route("AddCourse/Summary")]
+        public IActionResult? Summary(SummaryViewModel model)
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>()!;
+
+            try
+            {
+                var centreId = User.GetCentreId();

```

Comment:
I feel like the contents of this try should perhaps be a service method with a transaction as well, but there is quite a bit of mapping with models that aren't from the data project here

Reply:
should I leave it as is then?